### PR TITLE
add json to metadata response side support

### PR DESCRIPTION
**Pull request by cqi1217**
Commit Message:
Currently the json to metadata filter can only extract value from request body, this PR adds the support to extract value from response body.
Related PR: https://github.com/envoyproxy/envoy/pull/28394
Additional Description:
Risk Level:low
Testing:unit tests
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto [hunk 1]
```diff
@@ -112,5 +112,8 @@ message JsonToMetadata {
   }
 
   // Rules to match json body of requests
-  MatchRules request_rules = 1 [(validate.rules).message = {required: true}];
+  MatchRules request_rules = 1;
+
+  // Rules to match json body of responses
+  MatchRules response_rules = 2;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:118]: Please add a clarification that at least one of `request_rules` and `response_rules` must be provided.

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a couple of minor API comments.
Assigning @JuniorHsu as code-owner.
/assign @JuniorHsu

File: source/extensions/filters/http/json_to_metadata/filter.cc [hunk 2]
```diff
@@ -189,42 +203,49 @@ void Filter::finalizeDynamicMetadata(Http::StreamFilterCallbacks& filter_callbac
   }
```
Comment (JuniorHsu) [inline source/extensions/filters/http/json_to_metadata/filter.cc:203]: we should do this in request_only given route is not changed after router filter in request.

File: source/extensions/filters/http/json_to_metadata/filter.cc [hunk 3]
```diff
@@ -177,54 +191,68 @@ bool Filter::addMetadata(const std::string& meta_namespace, const std::string& k
 }
 
 void Filter::finalizeDynamicMetadata(Http::StreamFilterCallbacks& filter_callback,
-                                     const StructMap& struct_map, bool& processing_finished_flag) {
+                                     const StructMap& struct_map, bool& processing_finished_flag,
+                                     bool shouldClearRouteCache) {
```
Comment (JuniorHsu) [inline source/extensions/filters/http/json_to_metadata/filter.cc:195]: should_clear_route_cache here and there

File: source/extensions/filters/http/json_to_metadata/filter.cc [hunk 4]
```diff
@@ -86,30 +86,36 @@ FilterConfig::FilterConfig(
...
+      request_rules_(generateRules(proto_config.request_rules().rules())),
+      response_rules_(generateRules(proto_config.response_rules().rules())),
+      request_allow_content_types_(
+          generateAllowContentTypes(proto_config.request_rules().allow_content_types())),
+      response_allow_content_types_(
+          generateAllowContentTypes(proto_config.response_rules().allow_content_types())),
+      request_allow_empty_content_type_(proto_config.request_rules().allow_empty_content_type()),
+      response_allow_empty_content_type_(proto_config.response_rules().allow_empty_content_type()) {
+  if (request_rules_.empty() && response_rules_.empty()) {
+    throw EnvoyException("json_to_metadata_filter: Per filter configs must at least specify "
```
Comment (JuniorHsu) [inline source/extensions/filters/http/json_to_metadata/filter.cc:98]: let have a new config_test to test this?

File: test/extensions/filters/http/json_to_metadata/filter_test.cc [hunk 5]
```diff
@@ -104,15 +122,27 @@ class FilterTest : public testing::Test {
...
+  NiceMock<Http::MockStreamEncoderFilterCallbacks> encoder_callbacks_;
   NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info_;
   envoy::config::core::v3::Metadata dynamic_metadata_;
   std::shared_ptr<FilterConfig> config_;
   std::shared_ptr<Filter> filter_;
   Buffer::OwnedImpl buffer_;
   Http::TestRequestHeaderMapImpl incoming_headers_{
       {":path", "/ping"}, {":method", "GET"}, {"Content-Type", "application/json"}};
+  Http::TestResponseHeaderMapImpl response_headers_{
+      {":path", "/ping"}, {":method", "GET"}, {"Content-Type", "application/json"}};
```
Comment (JuniorHsu) [inline test/extensions/filters/http/json_to_metadata/filter_test.cc:145]: let's remove path/method and add status 200 to match common sense

File: test/extensions/filters/http/json_to_metadata/filter_test.cc [hunk 6]
```diff
@@ -143,6 +173,30 @@ TEST_F(FilterTest, BasicStringMatch) {
   EXPECT_EQ(getCounterValue("json_to_metadata.rq_invalid_json_body"), 0);
 }
 
+TEST_F(FilterTest, BasicResponseStringMatch) {
+  initializeFilter(request_config_yaml_);
```
Comment (JuniorHsu) [inline test/extensions/filters/http/json_to_metadata/filter_test.cc:177]: rename to config_yaml_

**Review by JuniorHsu — CHANGES_REQUESTED** [review]

**Comment by JuniorHsu** [comment]
try merge the main to see if ci passes?

**Comment by cqi1217** [comment]
seems it is still failing

File: source/extensions/filters/http/json_to_metadata/filter.h [hunk 7]
```diff
@@ -94,37 +101,52 @@ class Filter : public Http::PassThroughFilter, Logger::Loggable<Logger::Id::filt
...
   // the value of `parent_node->key`.
   absl::Status handleOnPresent(Json::ObjectSharedPtr parent_node, const std::string& key,
-                               const Rule& rule, StructMap& struct_map);
+                               const Rule& rule, StructMap& struct_map,
+                               Http::StreamFilterCallbacks& filter_callback);
 
   // Process the case without body, i.e., on_missing is applied for all rules.
-  void handleAllOnMissing(const Rules& rules, bool& processing_finished_flag);
+  void handleAllOnMissing(const Rules& rules, bool& processing_finished_flag,
+                          Http::StreamFilterCallbacks& filter_callback, bool shouldClearRouteCache);
```
Comment (JuniorHsu) [inline source/extensions/filters/http/json_to_metadata/filter.h:122]: snake style in *.h file

also input variable should in front of output variable. `processing_finished_flag` and `filter_callback` are probably need to be after should_clear_route_cache
https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

**Review by JuniorHsu — COMMENTED** [review]
only one nit

**Review by JuniorHsu — APPROVED** [review]
over to maintainer's review @KBaichoo

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by cqi1217** [comment]
@KBaichoo could you take a look? thanks

**Comment by alyssawilk** [comment]
sorry this wasn't assigned to Kevin so wouldn't show up on his PR reminders - fixed that today

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto [hunk 8]
```diff
@@ -111,6 +111,10 @@ message JsonToMetadata {
     bool allow_empty_content_type = 3;
   }
 
-  // Rules to match json body of requests
-  MatchRules request_rules = 1 [(validate.rules).message = {required: true}];
+  // At lease one of request_rules and response_rules must be provided.
```
Comment (KBaichoo) [inline api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:114]: s/lease/least/

File: source/extensions/filters/http/json_to_metadata/filter.h [hunk 9]
```diff
@@ -94,37 +101,54 @@ class Filter : public Http::PassThroughFilter, Logger::Loggable<Logger::Id::filt
...
   // Process the case with error, i.e., on_error is applied for all rules.
-  void handleAllOnError(const Rules& rules, bool& processing_finished_flag);
+  void handleAllOnError(const Rules& rules, bool should_clear_route_cache,
+                        Http::StreamFilterCallbacks& filter_callback,
+                        bool& processing_finished_flag);
   // Parse the body while we have the whole json.
-  void processBody(const Buffer::Instance* body, const Rules& rules, bool& processing_finished_flag,
-                   Stats::Counter& success, Stats::Counter& no_body, Stats::Counter& non_json);
+  void processBody(const Buffer::Instance* body, const Rules& rules, bool should_clear_route_cache,
+                   Stats::Counter& success, Stats::Counter& no_body, Stats::Counter& non_json,
```
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:130]: nit: this method has lots of args, consider encapping the various stat references in a wrapper struct.

**Review by KBaichoo — COMMENTED** [review]
Sorry for the delay, this otherwise looks good. Thank you for working on this.

/wait

**Comment by cqi1217** [comment]
addressed comments

**Review by KBaichoo — APPROVED** [review]
Thank you!

**Pull request merged**